### PR TITLE
Add ReadPeak Bid Adapter

### PR DESCRIPTION
**Pull request by kurrestahlberg**
## Type of change

- [X] New bidder adapter   

## Description of change

Bid adapter for readpeak.com, a purely native bidder from Finland

- test parameters for validating bids
```javascript
{
  bidder: 'readpeak',
  params: {
     bidfloor: 5.00,
     publisherId: 'test'
   },
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer kurre.stahlberg@readpeak.com
- [X] official adapter submission

### CODE REVIEW

File: modules/readpeakBidAdapter.js [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    }));
+  }
+  const bids = [];
+  Object.keys(idToImpMap).forEach(id => {
+    if (idToBidMap[id]) {
+      const bid = {
+        requestId: id,
+        cpm: idToBidMap[id].price,
+        creativeId: id,
+        adId: id,
```
Comment (mkendall07) [inline modules/readpeakBidAdapter.js]: what's `adId` ? We don't require that param.

File: modules/readpeakBidAdapter.js [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+      idToBidMap[bid.impid] = bid;
+    }));
+  }
+  const bids = [];
+  Object.keys(idToImpMap).forEach(id => {
+    if (idToBidMap[id]) {
+      const bid = {
+        requestId: id,
+        cpm: idToBidMap[id].price,
+        creativeId: id,
```
Comment (mkendall07) [inline modules/readpeakBidAdapter.js]: Is the creativeId really just the request id? Doesn't seem correct.

File: modules/readpeakBidAdapter.js [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        adId: id,
+        ttl: 300,
+        netRevenue: true,
+        mediaType: 'native',
+        currency: bidResponse.cur,
+        bidderCode: BIDDER_CODE,
+      };
+      bid['native'] = nativeResponse(idToImpMap[id], idToBidMap[id]);
```
Comment (mkendall07) [inline modules/readpeakBidAdapter.js]: no need to use bracket notation here.

File: modules/readpeakBidAdapter.js (hunk 1 above)
Comment (mkendall07) [inline modules/readpeakBidAdapter.js]: `adId` can be dropped. I forgot this was the legacy bid identifier. It will be found on the `requestId` now

**Comment by mkendall07** [comment]
@kurrestahlberg 
Thanks for the submission! Needs a few minor updates. 

Also it looks like I'm not receiving bid response. Here is the request generated to assist with testing:
```
{"id":"35bb8c9a63e32c","imp":[{"id":"486443e023cc9e","native":{"request":"{\"assets\":[{\"id\":1,\"required\":1,\"title\":{\"len\":70}},{\"id\":4,\"required\":1,\"data\":{\"type\":2,\"len\":120}}]}","ver":"1.1"},"bidfloor":5,"bidfloorcur":"USD"}],"site":{"publisher":{"id":"test"},"id":"test","ref":"","page":"http://ap.localhost:9999/integrationExamples/native_test.html","domain":"ap.localhost"},"app":null,"device":{"ua":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.94 Safari/537.36","language":"en-US"},"isPrebid":true}
```

**Comment by kurrestahlberg** [comment]
Thanks for the review! I have fixed the issues you pointed out. We seem to have a minor configuration issue in our backend right now and thus it returns 204 for the test publisher id. I will get it fixed asap.

**Comment by kurrestahlberg** [comment]
@mkendall07 
Everything should be in order again. Please retest.

File: modules/readpeakBidAdapter.js [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+    if (idToBidMap[id]) {
+      const bid = {
+        requestId: id,
+        cpm: idToBidMap[id].price,
+        creativeId: idToBidMap[id].crid,
+        ttl: 300,
+        netRevenue: true,
+        mediaType: 'native',
+        currency: bidResponse.cur,
+        bidderCode: BIDDER_CODE,
```
Comment (mkendall07) [inline modules/readpeakBidAdapter.js]: please drop bidder code. This will be added automatically.

File: modules/readpeakBidAdapter.js [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+    return {
+      method: 'POST',
+      url: ENDPOINT,
+      data: JSON.stringify(request),
+    }
+  },
+
+  interpretResponse: (response, request) => (
+    bidResponseAvailable(request, response)
```
Comment (mkendall07) [inline modules/readpeakBidAdapter.js]: Missing a return statement here.

**Comment by mkendall07** [comment]
@kurrestahlberg 
if you make those changes quick I'll retest

**Comment by mkendall07** [comment]
seeing bids now. Just need `bidderCode` dropped and we are good to merge

**Comment by kurrestahlberg** [comment]
@mkendall07 
Should be okay now. Sorry for the extra commits, the tests run so slow on my laptop that I always push while they are running...

**Comment by kurrestahlberg** [comment]
I don't know how to trigger rerun for travis.

**Comment by mkendall07** [comment]
restarted the build

**Comment by mkendall07** [comment]
@kurrestahlberg 
I'm seeing bid returned but our test creative isn't displaying. I haven't had a chance to dig further into it. Do you have a test page that is working?

**Comment by kurrestahlberg** [comment]
@mkendall07 
I'm having trouble getting anything to show here, but I can see that the native parts are sent and they look correct to me.

The following is sent as prev_scp:
```
hb_native_linkurl=https%253A%252F%252Fapp.readpeak.com%252Fads%252Ftracker%252F528%252F4_askelta_vaikuttavaan_natiivimainontaan%253Fl%253D95%2526a%253D3149%2526ka%253D1534%2526avn%253D1%2526b%253D487b0c08-74e5-49d4-8e48-9ad46630f622&hb_native_cta=&hb_native_image=https%3A%2F%2Fs3.eu-central-1.amazonaws.com%2Freadpeak%2Ffiles%2Farticle_thumbs%2Fthumb-408-c38808e3de9412fde8ce40b7a652f8cb.jpg&hb_native_brand=&hb_native_body=ReadPeak-k%C3%A4ytt%C3%B6liittym%C3%A4%20on%20kevyt%20ja%20ketter%C3%A4%20natiivimainonnan%20ty%C3%B6kalu%20jonka%20kautta%20tavoitat%20aiheistasi%20kiinnostun...&hb_native_title=4%20askelta%20vaikuttavaan%20natiivimainontaan&hb_size=0x0&hb_pb=5.80&hb_adid=24c0f8e5758435&hb_bidder=readpeak
```

The only odd bit to me is hb_size, and I am not sure if the regular test page will show both banners and native.

**Comment by mkendall07** [comment]
Ok figured it out - I wasn't overidding the bid CPM and our native placement only responds to nice round numbers...

anyway LGTM: 
image

**Pull request merged**